hilti/toolchain/include/compiler/detail/cfg.h

```diff
@@ -0,0 +1,247 @@
+    MetaNode(node::Tags node_tags) : Node(nullptr, node_tags, {}, {}) {}
+#pragma GCC diagnostic pop
+
+    HILTI_NODE_0(MetaNode, override);
+};
+
+/**
+ * A meta node for the start of a control flow.
+ */
+struct Start : MetaNode {
+    Start() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(Start, MetaNode, final);
+};
+
+/**
+ * A meta node joining or splitting control flow with no matching source statement.
+ */
+struct Flow : MetaNode {
+    Flow() : MetaNode(NodeTags) {}
+    HILTI_NODE_1(Flow, MetaNode, final);
+};
+
+/**
+ * A meta node to signify end of a scope carrying the source range of that scope.
+ */
+struct End : MetaNode {
+    End(const Node* scope) : MetaNode(NodeTags), scope(scope) {
+        assert(scope); // Should always contain a valid scope.
+    }
+
+    HILTI_NODE_1(End, MetaNode, final);
+
+    const Node* scope;
+};
+
+/**
+ * Node in a CFG.
+ *
+ * This class can deliberately only construct from a fixed set of AST nodes.
+ */
+class GraphNode {
+public:
+    GraphNode(operator_::function::Call* x) : node(x) {}
+    GraphNode(Expression* x) : node(x) {}
+    GraphNode(statement::Return* x) : node(x) {}
+    GraphNode(Statement* x) : node(x) {}
+    GraphNode(MetaNode* x) : node(x) {}
+    GraphNode(declaration::LocalVariable* x) : node(x) {}
+    GraphNode(declaration::GlobalVariable* x) : node(x) {}
+
+    GraphNode() = default;
+    GraphNode(const GraphNode&) = default;
+
+    GraphNode& operator=(const GraphNode& x) {
+        if ( &x == this )
+            return *this;
+
+        node = x.node;
+        return *this;
+    }
+
+    Node* operator->() { return node; }
+    const Node* operator->() const { return node; }
+
+    Node* value() const { return node; }
+
+    friend bool operator==(const GraphNode& a, const GraphNode& b) { return a.node == b.node; }
+    friend bool operator!=(const GraphNode& a, const GraphNode& b) { return ! (a.node == b.node); }
+
+    friend bool operator<(const GraphNode& a, const GraphNode& b) { return a.node < b.node; }
+
+private:
+    Node* node = nullptr;
+};
+
+} // namespace detail::cfg
+} // namespace hilti
+
+namespace std {
+template<>
+struct hash<hilti::detail::cfg::GraphNode> {
+    auto operator()(const hilti::detail::cfg::GraphNode& n) const { return n.value() ? n->identity() : 0; }
+};
+} // namespace std
+
+namespace hilti::detail::cfg {
+
+/**
+ * Dataflow facts about a node.
+ */
+struct Transfer {
+    /** Incoming edges, ordered by declaration they work on. */
+    std::unordered_map<Node*, std::unordered_set<GraphNode>> in;
+
+    /** Outgoing edges, ordered by declaration they work on. */
+    std::unordered_map<Node*, std::unordered_set<GraphNode>> out;
+
+    /** The previous nodes killed by this node. */
+    std::unordered_set<GraphNode> kill;
+
+    /** Set of declarations this node may alias. */
+    std::unordered_set<Declaration*> maybe_alias;
+
+    /** Values this node updates, ordered by declaration. */
```

Comment:
Can you clarify this comment? I don't quite get the meaning of key/value vs what is update what.

Reply:
I changed this to the comment below, let me know whether is helpful enough.

```
    /**
     * Declarations this graph node generates updates for.
     *
     * For each updated declaration we return the graph node were this value
     * was last updated.
     */
```